src/lib/storage/dictionary_segment.cpp

```diff
@@ -122,6 +123,207 @@ const ValueID DictionarySegment<T>::null_value_id() const {
+                             [&](const auto& row_id, const auto& search_value) { return !row_id.is_null(); })));
+      } else {
+        // TODO(cmfcmf): static_cast<uint32_t>(_null_value_id) is a workaround, because _null_value_id is a ValueID
+        // whereas the vector contains uints.
+        // Ideally, we should use some metaprogramming magic to get the type from the attribute_vector
+        non_null_end = static_cast<ChunkOffset>(std::distance(
+            attribute_vector.cbegin(),
+            std::lower_bound(
+                attribute_vector.cbegin(), attribute_vector.cend(), static_cast<uint32_t>(_null_value_id),
+                [&](const uint64_t value_id, const uint64_t null_value_id) { return value_id != null_value_id; })));
+      }
+    });
+  }
+
+  // std::cout << "non_null_end " << non_null_end << std::endl;
+
+  return non_null_end;
+}
+
+template <typename T>
+ChunkOffset DictionarySegment<T>::get_first_offset(const AllTypeVariant& search_value,
+                                                   const std::shared_ptr<const PosList>& position_filter) const {
+  Assert(_sort_order, "The segment needs to be sorted to calculate the first bound.");
+
+  const auto non_null_begin = get_non_null_begin_offset(position_filter);
+  const auto non_null_end = get_non_null_end_offset(position_filter);
+
+  const auto casted_search_value = type_cast_variant<T>(search_value);
+
+  ChunkOffset res = 0;
+
+  resolve_compressed_vector_type(*_attribute_vector, [&](const auto& attribute_vector) {
+    if (_sort_order.value() == OrderByMode::Ascending || _sort_order.value() == OrderByMode::AscendingNullsLast) {
+      if (position_filter != nullptr) {
+        // TODO(someone): if possible remove dictionary lookup in lambda function and just work on the value ids
+        const auto result =
+            std::lower_bound(position_filter->cbegin(), position_filter->cend(), casted_search_value,
+                             [&](const auto& row_id, const auto& search) {
+                               return _dictionary->operator[](_decompressor->get(row_id.chunk_offset)) < search;
+                             });
+        if (result == position_filter->cend()) {
+          res = INVALID_CHUNK_OFFSET;
+        } else {
+          res = static_cast<ChunkOffset>(std::distance(position_filter->cbegin(), result));
+        }
+      } else {
+        const auto begin = attribute_vector.cbegin() + non_null_begin;
+        const auto end = attribute_vector.cbegin() + non_null_end;
+        const auto result = std::lower_bound(
+            begin, end, casted_search_value,
+            [&](const auto& value_id, const auto& search) { return _dictionary->operator[](value_id) < search; });
+        if (result == end) {
+          res = INVALID_CHUNK_OFFSET;
+        } else {
+          res = static_cast<ChunkOffset>(std::distance(attribute_vector.cbegin(), result));
+        }
+      }
+    } else {
```

Comment:
The code in this else branch is a duplication of the code in the if branch except the compare lambda functions used for `std::lower_bound`.
You could remove the outer condition test `if (_sort_order.value() == OrderByMode::Ascending || _sort_order.value() == OrderByMode::AscendingNullsLast)` and pass different compare functions to `std::lower_bound`.

Reply:
👍